src/test/lib/ApiTest.cpp

```diff
@@ -1152,6 +1238,77 @@ DummyStreamCallback(
     return QUIC_STATUS_SUCCESS;
 }
 
+struct CloseFromCallbackContext {
+    uint16_t CloseCount;
+    uint16_t CurrentCount;
+
+    static QUIC_STATUS StreamCallback(_In_ MsQuicStream*, _In_opt_ void*, _Inout_ QUIC_STREAM_EVENT*) {
+        return QUIC_STATUS_SUCCESS;
+    }
+
+    static QUIC_STATUS Callback(_In_ MsQuicConnection* Conn, _In_opt_ void* Context, _Inout_ QUIC_CONNECTION_EVENT* Event, _In_ bool IsServer) {
+        QUIC_STATUS Status = QUIC_STATUS_SUCCESS;
+        auto Ctx = (CloseFromCallbackContext*)Context;
+
+        if (Event->Type == QUIC_CONNECTION_EVENT_PEER_STREAM_STARTED) {
+            new(std::nothrow) MsQuicStream(Event->PEER_STREAM_STARTED.Stream, CleanUpAutoDelete, StreamCallback, &IsServer);
+        }
+        if (Ctx->CloseCount == Ctx->CurrentCount++) {
+            Conn->Close();
+        }
+
+        return Status;
+    }
+
+    static QUIC_STATUS ClientCallback(_In_ MsQuicConnection* Conn, _In_opt_ void* Context, _Inout_ QUIC_CONNECTION_EVENT* Event) {
+        return Callback(Conn, Context, Event, false);
+    }
+    static QUIC_STATUS ServerCallback(_In_ MsQuicConnection* Conn, _In_opt_ void* Context, _Inout_ QUIC_CONNECTION_EVENT* Event) {
+        return Callback(Conn, Context, Event, true);
+    }
+};
+
+void
+QuicTestConnectionCloseFromCallback() {
+    for (int i = 0; i < 20; i++) {
+        CxPlatWatchdog Watchdog(2000);
+
+        CloseFromCallbackContext Context {(uint16_t)i, 0};
+
+        MsQuicRegistration Registration(true);
+        TEST_QUIC_SUCCEEDED(Registration.GetInitStatus());
+
+        MsQuicConfiguration ServerConfiguration(Registration, "MsQuicTest",
+            MsQuicSettings()
+                .SetPeerUnidiStreamCount(10)
+                .SetPeerBidiStreamCount(10)
+                .SetServerResumptionLevel(QUIC_SERVER_RESUME_AND_ZERORTT),
+            ServerSelfSignedCredConfig);
+        TEST_QUIC_SUCCEEDED(ServerConfiguration.GetInitStatus());
+
+        MsQuicConfiguration ClientConfiguration(Registration, "MsQuicTest",
+            MsQuicSettings()
+                .SetPeerUnidiStreamCount(10)
+                .SetPeerBidiStreamCount(10),
+        MsQuicCredentialConfig());
+        TEST_QUIC_SUCCEEDED(ClientConfiguration.GetInitStatus());
+
+        MsQuicAutoAcceptListener Listener(Registration, ServerConfiguration, CloseFromCallbackContext::ServerCallback, &Context);
```

Comment:
Right here is where `ServerCallback` is referenced. Below (line 1302) is where `ClientCallback` is referenced.